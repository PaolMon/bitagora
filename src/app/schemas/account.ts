export interface BitagoraAccountProps {
    infos: {
        myNeighbors: {
            address: Buffer,
            ipAddress: string,
            port: number
        }[],
        hostedFiles: {
            owner: Buffer,
            merkleRoot: Buffer
        }[],
    },
    digitalAsset: {
        pending: {
            fileName: string,
            merkleRoot: Buffer
        }[],
        allowed: {
            fileName: string,
            merkleRoot: Buffer,
            secret: string
        }[],
        myFiles: {
            fileName: string,
            merkleRoot: Buffer,
            secret:string
        }[],
    }
}

export const infosSchema = {
    type: 'object',
    properties: {
        myNeighbors: {
            fieldNumber: 1,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    address: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    ipAddress: {
                        dataType:'string',
                        fieldNumber: 2
                    },
                    port: {
                        dataType:'uint32',
                        fieldNumber: 3
                    }
                }
            }
        },
        hostedFiles: {
            fieldNumber: 2,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    owner: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 2
                    }
                }
            }
        }
    },
    default: {
        myNeighbors: [],
        hostedFiles: []
    }
}

export const digitalAssetAccountSchema = {
    type: 'object',
    properties: {
        myFiles: {
            fieldNumber: 1,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    fileName: {
                        dataType:'string',
                        fieldNumber: 2
                    },
                    secret: {
                        dataType:'string',
                        fieldNumber: 3
                    }
                }
            }
        },
        pending: {
            fieldNumber: 2,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    fileName: {
                        dataType:'string',
                        fieldNumber: 2
                    }
                }
            }
        },
        allowed: {
            fieldNumber: 3,
            type: 'array',
            items: {
                type: 'object',
                properties: {
                    merkleRoot: {
                        dataType:'bytes',
                        fieldNumber: 1
                    },
                    fileName: {
                        dataType:'string',
                        fieldNumber: 2
                    },
                    secret: {
                        dataType:'string',
                        fieldNumber: 3
                    }
                }
            }
        }
    },
    default: {
        myFiles: [],
        pending: [],
        allowed: []
    }
}
