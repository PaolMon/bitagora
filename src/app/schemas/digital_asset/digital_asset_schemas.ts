export const digitalAssetSchema = {
	$id: 'lisk/digital_asset/digitalAsset',
	type: 'object',
	required: ['owner', 'fileName', 'fileSize', 'fileHash', 'merkleRoot', 'merkleHeight', 'secret', 'transactionID', 'previousAssetReference'],
	properties: {
		owner: {
			dataType: 'bytes',
			fieldNumber: 1
		},
		fileName:{
			dataType: 'string',
			fieldNumber: 2
		},
		fileSize: {
			dataType: 'uint32',
			fieldNumber: 3
		},
		fileHash: {
			dataType: 'bytes',
			fieldNumber: 4
		},
		merkleRoot: {
			dataType: 'bytes',
			fieldNumber: 5
		},
		merkleHeight: {
			dataType: 'uint32',
			fieldNumber: 6
		},
		secret: {
			dataType: 'string',
			fieldNumber: 7
		},
		transactionID: {
			dataType: 'bytes',
			fieldNumber: 8
		},
		previousAssetReference: {
			dataType: 'bytes',
			fieldNumber: 9
		}
	},
};

export const registeredAssetsSchema = {
	$id: 'lisk/digital_asset/registeredAssets',
	type: "object",
	required: ["registeredAssets"],
	properties: {
	  registeredAssets: {
		type: "array",
		fieldNumber: 1,
		items: {
			type: 'object',
			required: ['owner', 'fileName', 'fileSize', 'fileHash', 'merkleRoot', 'merkleHeight', 'secret', 'transactionID', 'previousAssetReference'],
			properties: {
				owner: {
					dataType: 'bytes',
					fieldNumber: 1
				},
				fileName:{
					dataType: 'string',
					fieldNumber: 2
				},
				fileSize: {
					dataType: 'uint32',
					fieldNumber: 3
				},
				fileHash: {
					dataType: 'bytes',
					fieldNumber: 4
				},
				merkleRoot: {
					dataType: 'bytes',
					fieldNumber: 5
				},
				merkleHeight: {
					dataType: 'uint32',
					fieldNumber: 6
				},
				secret: {
					dataType: 'string',
					fieldNumber: 7
				},
				transactionID: {
					dataType: 'bytes',
					fieldNumber: 8
				},
				previousAssetReference: {
					dataType: 'bytes',
					fieldNumber: 9
				}
			},
		}
	  }
	}
	
};
