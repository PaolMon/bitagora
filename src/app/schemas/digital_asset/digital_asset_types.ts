export type digitalAsset = {
	owner:Buffer,
	fileName:string,
	fileSize:number,
	fileHash:Buffer,
	merkleRoot:Buffer,
	merkleHeight:number,
	secret:string,
	transactionID: Buffer,
	previousAssetReference: Buffer
}

export type registeredAssets = {
	registeredAssets: digitalAsset[]
}

export enum request_type {
	view_only = "VIEW",
	ownership = "OWN"
}

export enum request_status {
	pending = "PENDING",
	accepted = "ACCEPTED",
	rejected = "REJECTED"
}

export enum response_type {
	ok = "OK",
	ko = "KO"
}