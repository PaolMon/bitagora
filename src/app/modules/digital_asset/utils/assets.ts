import { BaseModuleDataAccess, StateStore } from "lisk-sdk";
import { codec } from "lisk-sdk";
import { digitalAsset, registeredAssets } from "../../../schemas/digital_asset/digital_asset_types";
import { registeredAssetsSchema } from "../../../schemas/digital_asset/digital_asset_schemas";

export const CHAIN_STATE_DIGITAL_ASSETS = "digitalAsset:registeredAssets";


export const setNewAsset= async (stateStore: StateStore, asset: digitalAsset) => {

    const DAs:digitalAsset[] = await getAllAssets(stateStore);

    DAs.push(asset);

    const registeredAssets: registeredAssets = {registeredAssets: DAs.sort((a, b) => a.merkleRoot.compare(b.merkleRoot))};

    await stateStore.chain.set(
        CHAIN_STATE_DIGITAL_ASSETS,
        codec.encode(registeredAssetsSchema, registeredAssets)
    );
}

export const getAllAssets = async (stateStore: StateStore) => {
    const registeredAssetsBuffer = await stateStore.chain.get(
        CHAIN_STATE_DIGITAL_ASSETS
    );

    if (!registeredAssetsBuffer) {
        return [];
    }

    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );

    return registeredAssets.registeredAssets;
}

export const getAssetByMerkleRoot = async (stateStore: StateStore, merkleRoot: Buffer) => {
    const DAs:digitalAsset[] = await getAllAssets(stateStore);

    const DA_index: number = DAs.findIndex((t) => t.merkleRoot.equals(merkleRoot));

    if (DA_index < 0) {
        throw new Error("Asset not found");
    }

    return DAs[DA_index];
}

/*
    THE FOLLOWING FUNCTIONS USE DATA ACCESS INSTEAD OF STATE STORE TO READ DATA
*/


export const _getAssetByMerkleRoot = async (dataAccess: BaseModuleDataAccess, merkleRoot: Buffer): Promise<digitalAsset> => {
    const registeredAssetsBuffer = await dataAccess.getChainState(
        CHAIN_STATE_DIGITAL_ASSETS
    );
    
    if (!registeredAssetsBuffer) {
        throw new Error('No Assets Found in stateStore');
    }
    
    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );
    const DAs = registeredAssets.registeredAssets;

    const DA_index: number = DAs.findIndex((t) => t.merkleRoot.equals(merkleRoot));

    if (DA_index < 0) {
        throw new Error("Asset not found");
    }

    return DAs[DA_index];
}

export const _getAllJSONAssets =async (dataAccess: BaseModuleDataAccess) => {
    const registeredAssetsBuffer = await dataAccess.getChainState(
        CHAIN_STATE_DIGITAL_ASSETS
    );
    
    if (!registeredAssetsBuffer) {
        return [];
    }
    
    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );
    
      return codec.toJSON(registeredAssetsSchema, registeredAssets);
}

export const _getAssetHistoryByMerkleRoot = async (dataAccess: BaseModuleDataAccess, merkleRoot: Buffer) => {

    const registeredAssetsBuffer = await dataAccess.getChainState(
        CHAIN_STATE_DIGITAL_ASSETS
    );
    
    if (!registeredAssetsBuffer) {
        return [];
    }
    
    const registeredAssets = codec.decode<registeredAssets>(
        registeredAssetsSchema,
        registeredAssetsBuffer
    );
    const DAs = registeredAssets.registeredAssets;

    let related_assets: digitalAsset[] = []
    let index = -1;
    let asset: digitalAsset;

    while(!merkleRoot.equals(Buffer.alloc(0))) {
        
        index = DAs.findIndex((t) => t.merkleRoot.equals(merkleRoot));

        if (index < 0) {
            throw new Error("Asset not found");
        }

        asset = DAs[index];
        related_assets.push(asset);

        merkleRoot = asset.previousAssetReference;
    }

    return codec.toJSON(registeredAssetsSchema, {registeredAssets: related_assets});
    //return related_assets;
}