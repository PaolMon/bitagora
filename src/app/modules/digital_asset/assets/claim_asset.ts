import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { digitalAsset } from '../../../schemas/digital_asset/digital_asset_types';
import { getAssetByMerkleRoot, setNewAsset } from '../utils/assets';
import { chunk } from '../../../schemas/chunks/chunk_types';
import { getChunkByMerkleRoot, setNewChunk } from '../utils/chunks';
import { BitagoraAccountProps } from '../../../schemas/account';


export type claim = {
	oldMerkleRoot: Buffer,
	newMerkleRoot: Buffer,
	newMerkleHeight: number,
	newHosts: Buffer[],
	newSecret: string
}

export class ClaimAsset extends BaseAsset {
	public name = 'claim';
  	public id = 3;

  	// Define schema for asset
	public schema = {
		$id: 'digitalAsset/claim-asset',
		title: 'ClaimAsset transaction asset for digitalAsset module',
		type: 'object',
		required: ['oldMerkleRoot','newMerkleRoot', 'newMerkleHeight', 'newHosts'], 
		properties: {
			oldMerkleRoot:{
				dataType: 'bytes',
				fieldNumber: 1
			},
			newMerkleRoot:{
				dataType: 'bytes',
				fieldNumber: 2
			},
			newMerkleHeight: {
				dataType: 'uint32',
				fieldNumber: 3
			},
			newHosts: {
				type: 'array',
				fieldNumber: 4,
				items: {
					dataType: 'bytes'
				}
			},
			newSecret: {
				dataType: 'string',
				fieldNumber: 5
			}
		},
	};

	public validate({ asset }: ValidateAssetContext<claim>): void {
		// Validate your asset
		if (asset.newMerkleHeight === 0 && asset.newHosts.length < 3) {
			throw new Error("If file has been splitted into chunks, it must be sent to at least 3 hosts");
		}
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<claim>): Promise<void> {
		
		const senderAddress: Buffer = transaction.senderAddress;
		const old_chunk: chunk = await getChunkByMerkleRoot(stateStore, asset.oldMerkleRoot);

		if (old_chunk.owner.compare(senderAddress) != 0) {
			throw new Error("You are not allowed to claim this asset");
		}

		const digital_asset: digitalAsset = await getAssetByMerkleRoot(stateStore, asset.oldMerkleRoot);

		digital_asset.owner = senderAddress;
		digital_asset.merkleRoot = asset.newMerkleRoot;
		digital_asset.merkleHeight = asset.newMerkleHeight;
		digital_asset.previousAssetReference = asset.oldMerkleRoot;
		digital_asset.transactionID = transaction.id;
		digital_asset.secret = asset.newSecret;
	
		const chunk: chunk = {
			owner: senderAddress,
			merkleRoot: asset.newMerkleRoot,
			hostedBy: [],
			requestedBy: [],
			allowedViewers: []
		} 
		
		const senderAccount = await stateStore.account.get<BitagoraAccountProps>(senderAddress);
		
		senderAccount.digitalAsset.myFiles.push({fileName: digital_asset.fileName, merkleRoot: digital_asset.merkleRoot, secret: asset.newSecret})

		await setNewChunk(stateStore, chunk);

		await setNewAsset(stateStore, digital_asset);

		await stateStore.account.set(senderAccount.address, senderAccount).catch();

	}
}
