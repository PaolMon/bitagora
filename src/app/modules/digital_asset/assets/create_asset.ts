import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { BitagoraAccountProps } from '../../../schemas/account';
import { chunk } from '../../../schemas/chunks/chunk_types';
import { digitalAsset } from '../../../schemas/digital_asset/digital_asset_types';
import { setNewAsset } from '../utils/assets';
import { checkChunkExistenceByMerkleRoot, setNewChunk } from '../utils/chunks';

export type create = {
	fileName:string,
	fileSize:number,
	fileHash:Buffer,
	merkleRoot:Buffer,
	merkleHeight:number,
	secret:string
}

export type my_file = {
	fileName: string, 
	merkleRoot: Buffer
}

export class CreateAsset extends BaseAsset {
	public name = 'create';
  	public id = 0;

  	// Define schema for asset
	public schema = {
    	$id: 'digitalAsset/create-asset',
		title: 'CreateAsset transaction asset for digitalAsset module',
		type: 'object',
		required: ['fileName', 'fileSize', 'fileHash', 'merkleRoot', 'merkleHeight', 'secret'],
		properties: {
			fileName:{
				dataType: 'string',
				fieldNumber: 1,
				maxLength: 64
			},
			fileSize: {
				dataType: 'uint32',
				fieldNumber: 2
			},
			fileHash: {
				dataType: 'bytes',
				fieldNumber: 3
			},
			merkleRoot: {
				dataType: 'bytes',
				fieldNumber: 4
			},
			merkleHeight: {
				dataType: 'uint32',
				fieldNumber: 5
			},
			secret: {
				dataType: 'string',
				fieldNumber: 6
			},
		},
  	};

	public validate({ asset }: ValidateAssetContext<create>): void {
		if (asset.fileName.length > 40) {
			throw new Error('file name longer than 40 char.')
		}
	}

	// eslint-disable-next-line @typescript-eslint/require-await
  	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<create>): Promise<void> {
	  	const new_digital_asset: digitalAsset = {
			owner: transaction.senderAddress,
			fileName: asset.fileName,
			fileSize: asset.fileSize,
			fileHash: asset.fileHash,
			merkleRoot: asset.merkleRoot,
			merkleHeight: asset.merkleHeight,
			secret: asset.secret,
			transactionID: transaction.id,
			previousAssetReference: Buffer.alloc(0)
		}
		
		const senderAddress = transaction.senderAddress;
		const senderAccount = await stateStore.account.get<BitagoraAccountProps>(senderAddress);
		
		senderAccount.digitalAsset.myFiles.push({fileName: new_digital_asset.fileName, merkleRoot: new_digital_asset.merkleRoot, secret: asset.secret})

		if(await checkChunkExistenceByMerkleRoot(stateStore, asset.merkleRoot)){
			throw new Error('this merkle root already exists in the stateStore')
		}
		
		const chunk: chunk = {
			owner: senderAddress,
			merkleRoot: asset.merkleRoot,
			hostedBy: [],
			requestedBy: [],
			allowedViewers: []
		} 

		await setNewChunk(stateStore, chunk);

		await setNewAsset(stateStore, new_digital_asset);

		await stateStore.account.set(senderAccount.address, senderAccount).catch();
		
	}
}