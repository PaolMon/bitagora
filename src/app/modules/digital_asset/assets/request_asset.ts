import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { BitagoraAccountProps } from '../../../schemas/account';
import { request_object } from '../../../schemas/chunks/chunk_types';
import { digitalAsset, request_status, request_type } from '../../../schemas/digital_asset/digital_asset_types';
import { getAssetByMerkleRoot } from '../utils/assets';
import { getChunkByMerkleRoot, updateChunk } from '../utils/chunks';

export type request = {
	merkleRoot: Buffer,
	mode: string
}

export class RequestAsset extends BaseAsset {
	public name = 'request';
	public id = 1;

	// Define schema for asset
		public schema = {
			$id: 'digitalAsset/request-asset',
			title: 'RequestAsset transaction asset for digitalAsset module',
			type: 'object',
			required: ['merkleRoot', 'mode'],
			properties: {
				merkleRoot:{
					dataType: 'bytes',
					fieldNumber: 1
				},
				mode: {
					dataType: 'string',
					fieldNumber: 2
				}
			}
	};

	public validate({ asset }: ValidateAssetContext<request>): void {
		// Validate your asset
		if(asset.mode !== request_type.ownership && asset.mode !== request_type.view_only) {
			throw new Error('Not a valid Request Type.')
		}
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<request>): Promise<void> {
		
		const senderAddress = transaction.senderAddress;
		
		let chunk = await getChunkByMerkleRoot(stateStore, asset.merkleRoot);

		if (senderAddress.compare(chunk.owner) === 0){
			throw new Error("Requester is already the owner of the Asset");
		}
		
		const index: number = chunk.requestedBy.findIndex((t) => (t.address.equals(senderAddress) && t.status !== request_status.rejected));
		if (index >= 0) {
			throw new Error("this address has already requested this asset");
		}

		let senderAccount = await stateStore.account.get<BitagoraAccountProps>(senderAddress);
		const index_a: number = senderAccount.digitalAsset.pending.findIndex((t) => t.merkleRoot.equals(asset.merkleRoot));
		if (index_a >= 0) {
			throw new Error("this Account has already requested this asset");
		}

		const ro: request_object = {
			address: senderAddress,
			requestTransaction: transaction.id,
			responseTransaction: Buffer.alloc(0),
			requestType: asset.mode,
			status: request_status.pending
		};
		chunk.requestedBy.push(ro);

		const da: digitalAsset = await getAssetByMerkleRoot(stateStore, asset.merkleRoot);
		senderAccount.digitalAsset.pending.push({
			fileName: da.fileName,
			merkleRoot: asset.merkleRoot
		});

		await updateChunk(stateStore, chunk);

		await stateStore.account.set(senderAccount.address, senderAccount).catch();
	}
}
