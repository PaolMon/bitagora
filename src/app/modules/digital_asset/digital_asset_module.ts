/*
 * LiskHQ/lisk-commander
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

/* eslint-disable class-methods-use-this */

import {
    AfterBlockApplyContext,


    AfterGenesisBlockApplyContext, BaseModule,


    BeforeBlockApplyContext, codec, TransactionApplyContext
} from 'lisk-sdk';
import { BitagoraAccountProps, digitalAssetAccountSchema } from '../../schemas/account';
import { digitalAssetSchema } from '../../schemas/digital_asset/digital_asset_schemas';
import { digitalAsset } from '../../schemas/digital_asset/digital_asset_types';
import { ClaimAsset } from "./assets/claim_asset";
import { CreateAsset } from "./assets/create_asset";
import { RequestAsset } from "./assets/request_asset";
import { ResponseAsset } from "./assets/response_asset";
import { _getAllJSONAssets, _getAssetByMerkleRoot, _getAssetHistoryByMerkleRoot } from './utils/assets';
import { _getAllJSONChunks } from './utils/chunks';

export class DigitalAssetModule extends BaseModule {
    public actions = {
        // Example below
        // getBalance: async (params) => this._dataAccess.account.get(params.address).token.balance,
        // getBlockByID: async (params) => this._dataAccess.blocks.get(params.id),
        // Example below
        // getBalance: async (params) => this._dataAccess.account.get(params.address).token.balance,
        // getBlockByID: async (params) => this._dataAccess.blocks.get(params.id),
        getAllAssets: async () => _getAllJSONAssets(this._dataAccess),
        getAllChunks: async () => _getAllJSONChunks(this._dataAccess),
        getAssetHistory:async (params: Record<string, unknown>) => {
            let { merkleRoot } = params;
            if (!Buffer.isBuffer(merkleRoot) && typeof merkleRoot === 'string') {
                merkleRoot = Buffer.from(merkleRoot, 'hex')
            } 
            return _getAssetHistoryByMerkleRoot(this._dataAccess, merkleRoot as Buffer);
        },

        getAsset:async (params:Record<string, unknown>): Promise<object> => {
            let { merkleRoot } = params;
            if (!Buffer.isBuffer(merkleRoot) && typeof merkleRoot === 'string') {
                merkleRoot = Buffer.from(merkleRoot, 'hex')
            } 
            const asset = await _getAssetByMerkleRoot(this._dataAccess, merkleRoot as Buffer);
            //return JSON.stringify(asset);
            return codec.toJSON(digitalAssetSchema, asset)
        },

        getAssetOwner: async(params: Record<string, unknown>): Promise<string> => {
            let { merkleRoot } = params;
            if (!Buffer.isBuffer(merkleRoot) && typeof merkleRoot === 'string') {
                merkleRoot = Buffer.from(merkleRoot, 'hex')
            } 
            const asset:digitalAsset = await _getAssetByMerkleRoot(this._dataAccess, merkleRoot as Buffer);
            return asset.owner.toString('hex');
        },

        getAccountAssets:async (params:Record<string, unknown>) => {
            let { address } = params;
            if (!Buffer.isBuffer(address) && typeof address === 'string') {
                address = Buffer.from(address, 'hex')
            } 
            const account = await this._dataAccess.getAccountByAddress<BitagoraAccountProps>(address as Buffer);
            return account.digitalAsset;
        },
    };
    public reducers = {
        // Example below
        // getBalance: async (
		// 	params: Record<string, unknown>,
		// 	stateStore: StateStore,
		// ): Promise<bigint> => {
		// 	const { address } = params;
		// 	if (!Buffer.isBuffer(address)) {
		// 		throw new Error('Address must be a buffer');
		// 	}
		// 	const account = await stateStore.account.getOrDefault<TokenAccount>(address);
		// 	return account.token.balance;
		// },
    };
    public name = 'digitalAsset';
    public accountSchema = digitalAssetAccountSchema;
    public transactionAssets = [new CreateAsset(), new RequestAsset(), new ResponseAsset(), new ClaimAsset()];
    public events = [
        // Example below
        // 'digitalAsset:newBlock',
    ];
    public id = 1001;

    // public constructor(genesisConfig: GenesisConfig) {
    //     super(genesisConfig);
    // }

    // Lifecycle hooks
    public async beforeBlockApply(_input: BeforeBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async afterBlockApply(_input: AfterBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async beforeTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterGenesisBlockApply(_input: AfterGenesisBlockApplyContext) {
        // Get any data from genesis block, for example get all genesis accounts
        // const genesisAccoounts = genesisBlock.header.asset.accounts;
    }
}
