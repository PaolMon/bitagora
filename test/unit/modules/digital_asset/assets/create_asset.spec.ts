import { Account } from '@liskhq/lisk-chain';
import { ApplyAssetContext, codec, StateStore, testing } from 'lisk-sdk';
import { CHAIN_STATE_DIGITAL_ASSETS } from '../../../../../src/app/modules/digital_asset/utils/assets';
import { create, CreateAsset } from '../../../../../src/app/modules/digital_asset/assets/create_asset';
import { DigitalAssetModule } from '../../../../../src/app/modules/digital_asset/digital_asset_module';
import { BitagoraAccountProps } from '../../../../../src/app/schemas/account';
import { digitalAsset } from '../../../../../src/app/schemas/digital_asset/digital_asset_types';
import { registeredAssetsSchema } from '../../../../../src/app/schemas/digital_asset/digital_asset_schemas';

describe('CreateAsset', () => {
  let transactionAsset: CreateAsset;

	beforeEach(() => {
		transactionAsset = new CreateAsset();
	});

	describe('constructor', () => {
		it('should have valid id', () => {
			expect(transactionAsset.id).toEqual(0);
		});

		it('should have valid name', () => {
			expect(transactionAsset.name).toEqual('create');
		});

		it('should have valid schema', () => {
			expect(transactionAsset.schema).toMatchSnapshot();
		});
	});

	describe('validate', () => {
		describe('schema validation', () => {
			it.todo('should throw errors for invalid schema');
			it('should be ok for valid schema', () => {
				const create_object: create = {
					fileName: 'prova.txt',
					fileSize: 3,
					fileHash: Buffer.alloc(5),
					merkleRoot: Buffer.alloc(6),
					merkleHeight: 0,
					secret: 'prova'
				};
				const context = testing.createValidateAssetContext({
					asset: create_object,
					transaction: { senderAddress: Buffer.alloc(0) } as any,
				});

				expect(() => transactionAsset.validate(context)).not.toThrow();
			});
    	});
	});

	describe('apply', () => {
		let stateStore: StateStore;
		let account: Account<BitagoraAccountProps>;
		let context: ApplyAssetContext<create>;

		beforeEach(() => {
			account = testing.fixtures.createDefaultAccount<BitagoraAccountProps>([DigitalAssetModule]);
	
			stateStore = new testing.mocks.StateStoreMock({
				accounts: [account]
			});

			const create_object: create = {
				fileName: 'prova.txt',
				fileSize: 3,
				fileHash: Buffer.alloc(4, 'hash'),
				merkleRoot: Buffer.alloc(10, 'merkleroot'),
				merkleHeight: 0,
				secret: 'prova'
			}
	
			context = testing.createApplyAssetContext<create>({
				stateStore,
				asset: create_object,
				transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
			});
	
			jest.spyOn(stateStore.chain, 'get');
			jest.spyOn(stateStore.chain, 'set');
		});


    	describe('valid cases', () => {
      		it('should update the state store',async () => {
				await transactionAsset.apply(context);
				const new_digital_asset: digitalAsset = {
					owner: context.transaction.senderAddress,
					fileName: 'prova.txt',
					fileSize: 3,
					fileHash: Buffer.alloc(4, 'hash'),
					merkleRoot: Buffer.alloc(10, 'merkleroot'),
					merkleHeight: 0,
					secret: 'prova',
					transactionID: context.transaction.id,
					previousAssetReference: Buffer.alloc(0)
				}
				const DAs: digitalAsset[] = []
				DAs.push(new_digital_asset)		
				expect(stateStore.chain.set).toHaveBeenCalledWith(
					CHAIN_STATE_DIGITAL_ASSETS,
					codec.encode(registeredAssetsSchema, {registeredAssets:DAs})
				);
			});
			it('should update sender account',async () => {
				await transactionAsset.apply(context);
				const senderAccount = await stateStore.account.get<BitagoraAccountProps>(context.transaction.senderAddress)	
				expect(senderAccount.digitalAsset.myFiles).toContainEqual({
						fileName: context.asset.fileName, 
						merkleRoot: context.asset.merkleRoot, 
						secret: context.asset.secret
				})
		  	});
    	});

    	describe('invalid cases', () => {
      		it.todo('should throw error');
    	});
	});
});
