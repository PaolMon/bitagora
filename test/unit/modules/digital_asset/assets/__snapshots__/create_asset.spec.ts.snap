// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAsset constructor should have valid schema 1`] = `
Object {
  "$id": "digitalAsset/create-asset",
  "properties": Object {
    "fileHash": Object {
      "dataType": "bytes",
      "fieldNumber": 3,
    },
    "fileName": Object {
      "dataType": "string",
      "fieldNumber": 1,
      "maxLength": 64,
    },
    "fileSize": Object {
      "dataType": "uint32",
      "fieldNumber": 2,
    },
    "merkleHeight": Object {
      "dataType": "uint32",
      "fieldNumber": 5,
    },
    "merkleRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 4,
    },
    "secret": Object {
      "dataType": "string",
      "fieldNumber": 6,
    },
  },
  "required": Array [
    "fileName",
    "fileSize",
    "fileHash",
    "merkleRoot",
    "merkleHeight",
    "secret",
  ],
  "title": "CreateAsset transaction asset for digitalAsset module",
  "type": "object",
}
`;
